### ISSUE TITLE
Curious train loss graph for bidaf training 

### ISSUE THREAD

**Issue by aman313** [allenai/allennlp#1706]
![image](https://user-images.githubusercontent.com/2459283/45028461-bed95d80-b062-11e8-9ac5-d26b50afe7bd.png)

Hi folks,
I noticed the above loss graph while training bidaf on squad data. Curiously, it looks like there is a periodic nature to the fall in training loss which corresponds to the end of an epoch. Even more weirdly the training loss seems to be increasing slightly during the epoch. I saw this while using version 0.5.1 and also 0.6.1. I could not find any obvious bugs in the code but I am not sure what would cause such a behavior. Thoughts?

**Comment by aman313** [comment]
Hey Matt. I did see that the loss is averaged over batches. But I am still not sure how an average increasing (or even remaining constant) over an epoch would result in a lower loss the next epoch. Thanks.

**Comment by matt-gardner** [comment]
It depends on the order that you see your instances.  Notice that in your graph, the loss doesn't increase over every epoch.  It's just random based on which instances you happened to get early on in the epoch.  If you look at the average loss at the end of each epoch, you see a steadily decreasing curve.

**Comment by sethah** [comment]
This seems like a weird design choice to me. Taking the running average with infinite history per epoch produces this weird stepped signal, as below, which badly overestimates the loss in early training.

![image](https://user-images.githubusercontent.com/7275795/50623661-05ce4700-0ecd-11e9-9786-e900535c48bb.png)

I would vote to change this to showing the average over the last `summary_interval` iterations. I was initially quite confused by the stepping behavior and wondered if something strange was happening in my model training. I had to look at the source code before it made sense. 

If such a change is supported, I'd be happy to submit a PR.

**Comment by matt-gardner** [comment]
No objections to this change from me, I think what you suggest is better than what we have.

**Comment by sethah** [comment]
Cool, I'll work on it.
